servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,335 @@
+            ByteBuf origin = toByteBuf(src, offset, length);
+            channel.writeInbound(origin);
+            return newBufferFrom(channel.readInbound());
+        } catch (Throwable e) {
+            LOGGER.error("Error while decoding with {}", name(), e);
+            throwException(e);
+            return EMPTY_BUFFER;
+        } finally {
+            safeCleanup(channel);
+        }
+    }
+
+    @Override
+    public Publisher<Buffer> decode(final Publisher<Buffer> from, final BufferAllocator allocator) {
+        requireNonNull(from);
+        return from.liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+
+            private final ByteToMessageDecoder decoder = decoderSupplier.get();
+            private final EmbeddedChannel channel = newEmbeddedChannel(decoder, allocator);
+
+            @Nullable
+            PublisherSource.Subscription subscription;
+
+            @Override
+            public void onSubscribe(final PublisherSource.Subscription subscription) {
+                this.subscription = subscription;
+                subscriber.onSubscribe(subscription);
+            }
+
+            @Override
+            public void onNext(@Nullable final Buffer src) {
+                assert subscription != null;
+
+                if (src == null) {
+                    return;
+                }
+
+                // onNext will produce AT-MOST N items (as received)
+                if (!channel.isOpen()) {
+                    throw new IllegalStateException("Stream encoder previously closed but more input arrived ");
+                }
+
+                channel.writeInbound(toByteBuf(src));
+                Buffer buffer = drainChannelInToSingleBuffer(channel, allocator);
+
+                if (buffer != null && buffer.readableBytes() > 0) {
+                    subscriber.onNext(buffer);
+                } else {
+                    // Not enough data to decompress, ask for more
+                    subscription.request(1);
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                safeCleanup(channel);
+                subscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                cleanup(channel);
+                subscriber.onComplete();
+            }
+        });
+    }
+
+    private EmbeddedChannel newEmbeddedChannel(final ChannelHandler handler, final BufferAllocator allocator) {
+        final EmbeddedChannel channel = new EmbeddedChannel(handler);
+        channel.config().setAllocator(getByteBufAllocator(allocator));
+        return channel;
+    }
+
+    private static ByteBuf toByteBuf(final Buffer buffer) {
+        return BufferUtils.toByteBuf(buffer);
+    }
+
+    private static ByteBuf toByteBuf(final Buffer buffer, final int offset, final int length) {
+        return BufferUtils.toByteBuf(buffer.readerIndex(buffer.readerIndex() + offset).readSlice(length));
```

Comment:
This method will throw if it's not a `NettyBuffer`. However, users can use read-only allocator for the payload body (sending a file as a `MappedByteBuffer` required to use read-only allocator). Consider using `extractByteBufOrCreate`

Reply:
🦅 👁️